cpp/src/arrow/compute/kernels/scalar_if_else.cc

```diff
@@ -1413,6 +1417,588 @@ struct CaseWhenFunctor<NullType> {
+        RETURN_NOT_OK(raw_builder->AppendNull());
+      }
+    }
+  }
+
+  ARROW_ASSIGN_OR_RAISE(auto temp_output, raw_builder->Finish());
+  *output = *temp_output->data();
+  return Status::OK();
+}
+
+template <typename Type>
+struct CaseWhenFunctor<Type, enable_if_base_binary<Type>> {
+  using offset_type = typename Type::offset_type;
+  using BuilderType = typename TypeTraits<Type>::BuilderType;
+  static Status Exec(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    if (batch[0].null_count() > 0) {
+      return Status::Invalid("cond struct must not have outer nulls");
+    }
+    if (batch[0].is_scalar()) {
+      return ExecVarWidthScalarCaseWhen(ctx, batch, out);
+    }
+    return ExecArray(ctx, batch, out);
+  }
+
+  static Status ExecArray(KernelContext* ctx, const ExecBatch& batch, Datum* out) {
+    return ExecVarWidthArrayCaseWhen(
+        ctx, batch, out,
+        // ReserveData
+        [&](ArrayBuilder* raw_builder) {
+          int64_t reservation = 0;
+          for (size_t arg = 1; arg < batch.values.size(); arg++) {
+            auto source = batch.values[arg];
+            if (source.is_scalar()) {
+              const auto& scalar =
+                  checked_cast<const BaseBinaryScalar&>(*source.scalar());
+              if (!scalar.value) continue;
+              reservation =
+                  std::max<int64_t>(reservation, batch.length * scalar.value->size());
+            } else {
+              const auto& array = *source.array();
+              const auto& offsets = array.GetValues<offset_type>(1);
+              reservation =
+                  std::max<int64_t>(reservation, offsets[array.length] - offsets[0]);
+            }
+          }
+          // checked_cast works since (Large)StringBuilder <: (Large)BinaryBuilder
+          return checked_cast<BuilderType*>(raw_builder)->ReserveData(reservation);
+        },
+        // AppendScalar
+        [](ArrayBuilder* raw_builder, const Scalar& raw_scalar) {
+          const auto& scalar = checked_cast<const BaseBinaryScalar&>(raw_scalar);
+          return checked_cast<BuilderType*>(raw_builder)
+              ->Append(scalar.value->data(),
+                       static_cast<offset_type>(scalar.value->size()));
+        },
+        // AppendArray
+        [](ArrayBuilder* raw_builder, const std::shared_ptr<ArrayData>& array,
+           const int64_t row) {
+          const offset_type* offsets = array->GetValues<offset_type>(1);
+          return checked_cast<BuilderType*>(raw_builder)
+              ->Append(array->buffers[2]->data() + offsets[row],
+                       offsets[row + 1] - offsets[row]);
+        });
+  }
+};
+
+// Given an array and a builder, append a slice of the array to the builder
+using ArrayAppenderFunc = std::function<Status(
+    ArrayBuilder*, const std::shared_ptr<ArrayData>&, int64_t, int64_t)>;
+
+static Status GetValueAppenders(const DataType& type, ArrayAppenderFunc* array_appender);
+
+struct GetAppenders {
```

Comment:
Hmm... this functionality may be useful elsewhere (and actually, perhaps it can already be reused in other places).

Why not add a virtual method to `ArrayBuilder` (e.g. `virtual AppendArraySlice(...)`) instead of keeping this private?

Reply:
I was debating this. I'll refactor this to be part of the builder instead.